File: src/main/java/no/nav/fo/forenkletdeploy/service/JiraIssueService.java
package no.nav.fo.forenkletdeploy.service;

import no.nav.fo.forenkletdeploy.domain.JiraIssue;
import no.nav.fo.forenkletdeploy.domain.JiraIssues;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.stereotype.Component;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

import static no.nav.fo.forenkletdeploy.util.Utils.withClient;

@Component
public class JiraIssueService {

    // https://jira.adeo.no/secure/RapidBoard.jspa?rapidView=11947
    private static final List<String> IKKE_INTERESSANTE_STATUSER = Arrays.asList(
            "Produktkø",
            "Analyse",
            "Verifisering",
            "Klar til utvikling",
            "Ferdig"
    );
    public static final String IKKE_INTERESSANTE_STATUSER_FORMATTERT = IKKE_INTERESSANTE_STATUSER.stream().map(s -> "'" + s + "'").collect(Collectors.joining(","));
    public static final String FO_JQL = "project=fo AND status not in (" + IKKE_INTERESSANTE_STATUSER_FORMATTERT + ")";

    @SuppressWarnings("unchecked")
    @Cacheable("jiraissues")
    public List<JiraIssue> getUserStories() {
        return withClient("https://jira.adeo.no/rest/api/2/search")
                .queryParam("jql", FO_JQL)
                .queryParam("maxResults", 500)
                .request()
                .get(JiraIssues.class)
                .issues;
    }

    @SuppressWarnings("unchecked")
    @Cacheable("jiraissue")
    public JiraIssue getUserStory(String issueId) {
        return withClient("https://jira.adeo.no/rest/api/2/issue/" + issueId)
                .request()
                .get(JiraIssue.class);
    }
}

File: src/main/kotlin/no/nav/fo/forenkletdeploy/service/JiraIssueService.kt
package no.nav.fo.forenkletdeploy.service

import com.github.javafaker.Faker
import no.nav.fo.forenkletdeploy.*
import org.springframework.stereotype.Service

import no.nav.fo.forenkletdeploy.util.Utils.withClient
import no.nav.fo.forenkletdeploy.util.stringToSeed
import org.springframework.cache.annotation.Cacheable
import java.util.*

interface IJiraIssueService {
    fun getUserStories(): List<JiraIssue>
    fun getUserStory(issueId: String): JiraIssue
}

open class JiraIssueService: IJiraIssueService {
    val IKKE_INTERESSANTE_STATUSER = arrayOf(
        "Analyse",
        "Verifisering",
        "Klar til utvikling",
        "Ferdig"
    ).joinToString(separator = ",")
    val FO_JQL = "project=fo AND status not in ($IKKE_INTERESSANTE_STATUSER)"

    @Cacheable("jiraissues")
    override fun getUserStories(): List<JiraIssue> =
            withClient("https://jira.adeo.no/rest/api/2/search")
                    .queryParam("jql", FO_JQL)
                    .queryParam("maxResults", 500)
                    .request()
                    .get(JiraIssues::class.java)
                    .issues

    @Cacheable("jiraissue")
    override fun getUserStory(issueId: String): JiraIssue =
            withClient("https://jira.adeo.no/rest/api/2/issue/$issueId")
                    .request()
                    .get(JiraIssue::class.java)
}

class MockJiraIssueService: IJiraIssueService {
    val STATUSKODER = arrayOf(
            "Under utvikling",
            "Fagfellevurdering",
            "Test 1",
            "Test 2"
    )

    override fun getUserStories(): List<JiraIssue> =
            (123..130).map { getUserStory("PUS-$it") }

    override fun getUserStory(issueId: String): JiraIssue {
        val faker = Faker(Random(stringToSeed(issueId)))
        return JiraIssue(
                id = issueId,
                key = issueId,
                fields = JiraIssueFields(
                        assignee = JiraIssuePerson(displayName = faker.pokemon().name()),
                        summary = faker.rickAndMorty().quote(),
                        status = JiraIssueStatus(
                                name = STATUSKODER[faker.number().numberBetween(0, STATUSKODER.size-1)]
                        )
                )
        )
    }

}